dask/dataframe/groupby.py

```diff
@@ -122,6 +126,185 @@ def _nunique_series_aggregate(df):
+    return _agg_finalize(stage2, finalizers)
+
+
+def _normalize_spec(spec, non_group_columns):
+    """
+    Return a list of ``(result_column, func, input_column)`` tuples.
+    """
+    if not isinstance(spec, dict):
+        spec = collections.OrderedDict(zip(non_group_columns, it.repeat(spec)))
+
+    res = []
+
+    if isinstance(spec, dict):
+        for input_column, subspec in spec.items():
+            if isinstance(subspec, dict):
+                res.extend(((input_column, result_column), func, input_column)
+                           for result_column, func in subspec.items())
+
+            else:
+                if not isinstance(subspec, list):
+                    subspec = [subspec]
+
+                res.extend(((input_column, funcname(func)), func, input_column)
+                           for func in subspec)
+
+    else:
+        raise ValueError("unsupported agg spec of type {}".format(type(spec)))
+
+    compounds = (list, tuple, dict)
+    use_flat_columns = not any(isinstance(subspec, compounds)
+                               for subspec in spec.values())
+
+    if use_flat_columns:
+        res = [(input_col, func, input_col) for (_, func, input_col) in res]
+
+    return res
+
+
+def _build_agg_args(spec):
+    """
+    Create the transformation functions for a normalized aggregate spec.
+
+    Returns
+    -------
+    chunk_funcs: a list of functions that are applied on grouped chunks of the
+        initial dataframe. Each function takes a grouped dataframe and should
+        return a dataframe of intermediate values with the group keys as an
+        index.
+
+    agg_funcs: a list of functions that are applied on the grouped dataframe
+        after all chunks have been concatinated. Each function takes a grouped
+        dataframe and should return a dataframe of intermediate values with the
+        group keys as an index.
+
+    finalizers: a list of result-column transform pairs, that are applied after
+        the ``agg_funcs``. They are used to create final results from
+        intermediate representations.
+    """
+    # generator for consecutive IDs (for intermediate results)
+    ids = it.count()
+    next_id = lambda prefix: '{}-{}'.format(prefix, next(ids))
+
+    simple_impl = {
+        'sum': (M.sum, M.sum),
+        'min': (M.min, M.min),
+        'max': (M.max, M.max),
+        'count': (M.count, M.sum),
+        'size': (M.size, M.sum),
+    }
+
+    known_np_funcs = {np.min: 'min', np.max: 'max'}
+
+    chunks = []
+    aggs = []
+    finalizers = []
+
+    for (result_column, func, input_column) in spec:
+        func = funcname(known_np_funcs.get(func, func))
+
+        if func in simple_impl.keys():
+            impls = _build_agg_args_simple(result_column, func, input_column,
+                                           next_id, simple_impl[func])
+
+        elif func == 'var':
+            raise NotImplementedError()
```

Comment:
The functions in `dask/dataframe/groupby.py` like `_var_chunk` may be of value here.